### PR TITLE
JBDS-3849 use promise to fetch requirements into cache folder

### PR DESCRIPTION
**Pull request by dgolovin**
simpler approach from Denis; use requirements folder instead of prefetch-dependencies; also re-enable creation of sha256 file in requirements folder as we'll need it later

add requirements-cache/ folder to gitignore

### CODE REVIEW

**Comment by dgolovin** [comment]
Should get back package-bundle task.

**Comment by dgolovin** [comment]
updated PR to restore package-bundle task

**Comment by dgolovin** [comment]
@ppitonak could you review?

**Comment by ppitonak** [comment]
- [x] `npm prune && rm -rf dist && rm -rf prefetch-dependencies && npm install`
- [x] `npm start`
- [x] `npm test`
- [ ] `npm run ui-test` **FAILED** but not a blocker for this PR
- [x] `npm run package-simple`
- [x] `(( $(stat -c %s dist/win/*installer.exe) > 36000000 )) || echo 'file too small'`
- [x] `(( $(stat -c %s dist/win/*installer.exe) < 40000000 )) || echo 'file too big'`
- [x] start the application
- [ ] `npm run package-bundle`
- [ ] `(( $(stat -c %s dist/win/*installer.exe) > 1500000000 )) || echo 'file too small'`
- [ ] `(( $(stat -c %s dist/win/*installer.exe) < 1600000000 )) || echo 'file too big'`
- [ ] start the application
- [ ] `echo "aaa" > prefetch-dependencies/cdk.zip && npm run package-bundle`
- [ ] `echo "aaa" > prefetch-dependencies/cdk.zip && npm run dist`
- [ ] `npm run package-simple`
- [ ] `(( $(stat -c %s dist/win/*installer.exe) > 36000000 )) || echo 'file too small'`
- [ ] `(( $(stat -c %s dist/win/*installer.exe) < 40000000 )) || echo 'file too big'`

**Comment by ppitonak** [comment]
@dgolovin task _prefetch_ in `npm run package-bundle` finished after 14ms which indicates that something is wrong or I have really fast internet connection ;)

File: gulpfile.js [hunk 1]
```diff
@@ -40,10 +41,15 @@ let zaExtra7z = path.join(zaRoot, '7z920_extra.7z');
...
 let rhExe = path.join(zaRoot, 'ResourceHacker.exe');
 let zaElectronPackage = path.join(zaRoot, artifactName + '-win32-x64');
-let bundled7z = path.join(zaRoot, artifactName +'-win32-x64.7z');
-let installerExe = resolveInstallerExePath('');
-
-gulp.task('transpile:app', function() {
+let bundled7z = path.join(zaRoot, artifactName +'-win32-x64.7z');
+let installerExe = resolveInstallerExePath('');
+
+process.on('uncaughtException', function(err) {
```
Comment (ppitonak) [inline gulpfile.js]: Any specific reason to have this? It would be still logged if we didn't have this handler, wouldn't it?

File: gulpfile.js [hunk 2]
```diff
@@ -120,28 +126,34 @@ gulp.task('download-7zip-extra', function() {
...
-gulp.task('prepare-tools', function(cb) {
+gulp.task('prepare-tools', function(cb) {
   runSequence(['download-7zip', 'download-7zip-extra', 'download-resource-hacker'],
     ['unzip-7zip', 'unzip-resource-hacker'], 'unzip-7zip-extra', cb);
 });
 
 // wrap electron-generated app to 7zip archive
 gulp.task('create-7zip-archive', function(cb) {
-  let packCmd = zaExe + ' a ' + bundled7z + ' ' + zaElectronPackage + path.sep + '*';
+  var packCmd = zaExe + ' a ' + bundled7z + ' ' + zaElectronPackage + path.sep + '*'
```
Comment (ppitonak) [inline gulpfile.js]: Why did you change `let` to `var`?

File: gulpfile.js [hunk 3]
```diff
@@ -269,40 +272,46 @@ function resolveInstallerExePath(artifactType) {
...
 }
 
-// download all the installer dependencies so we can package them up into the .exe
-gulp.task('prefetch', function(cb) {
-  let counter=0;
+// prefetch all the installer dependencies so we can package them up into the .exe
+gulp.task('prefetch', function() {
+  let promises = new Set();
+  if (!fs.existsSync(prefetchFolder)) {
+     mkdirp(prefetchFolder);
```
Comment (ppitonak) [inline gulpfile.js]: mkdirp is an async function, it doesn't cause probles at this point but should be moved to separate gulp task similarly to _create-dist-win-dir_

File: gulpfile.js (hunk 2 above)
Comment (nickboldt) [inline gulpfile.js]: http://stackoverflow.com/questions/762011/let-keyword-vs-var-keyword

File: gulpfile.js [hunk 4]
```diff
@@ -269,40 +272,46 @@ function resolveInstallerExePath(artifactType) {
...
-        }
-      });
+          resolve(true);
+        });
+      }));
     }
-  }
-  installerExe = resolveInstallerExePath('-bundle');
+  }
+  return Promise.all(promises).then((result)=>{
```
Comment (nickboldt) [inline gulpfile.js]: Promise.all completes really fast even though the async downloads are still running.

That's why I gave up on using promises here and went back to the counter++ variable in #158

File: gulpfile.js (hunk 2 above)
Comment (ppitonak) [inline gulpfile.js]: Can you be more specific, please? `let` doesn't influence variable's visibility inside `if` statement and we don't need to use the variable before it's declared.

File: gulpfile.js [hunk 5]
```diff
@@ -269,40 +272,46 @@ function resolveInstallerExePath(artifactType) {
...
+                  return resolve(res);
                 }
               });
+            }
           }
-        } else {
-          console.log('DOWNLOADED <- ' + currentFile);
-        }
-      });
+          resolve(true);
```
Comment (jrichter1) [inline gulpfile.js]: This resolve is getting called right after download is initiated, thus prefetch finishes in a few milliseconds. You need to put it into an else block because of the magic way promises are handled.

File: gulpfile.js [hunk 6]
```diff
@@ -269,40 +272,46 @@ function resolveInstallerExePath(artifactType) {
...
+            if (reqs[currentKey].bundle === 'yes') {
+              console.log('[INFO] Download ' + currentUrl + ' to ' + currentFile);
+              request(currentUrl).pipe(fs.createWriteStream(currentFile)).on('finish',function(err, res) {
+                if (err) {
+                  console.log(error);
+                  return reject(err);
+                } else {
+                  //console.log('[DEBUG] Downloaded ' + currentFile);
+                  createSHA256File(currentFile);
+                  return resolve(res);
```
Comment (ppitonak) [inline gulpfile.js]: modify to
```
createSHA256File(currentFile, () => {
  return resolve();
});
```

File: gulpfile.js [hunk 7]
```diff
@@ -269,40 +272,46 @@ function resolveInstallerExePath(artifactType) {
...
-            console.log('DOWNLOADING -> ' + currentUrl);
-            request(currentUrl)
-              .pipe(fs.createWriteStream(currentFile)).on('finish',function() {
-                // create a sha256sum file
-                counter--;
-                if(counter===0) {
-                  cb();
+        isExistingSHA256Current(currentFile, reqs[currentKey].sha256sum, (alreadyDownloaded)=> {
+          //console.log('For ' + currentFile + ": " + alreadyDownloaded + ", " + reqs[currentKey].bundle);
+          if(!alreadyDownloaded) {
```
Comment (ppitonak) [inline gulpfile.js]: this branch is missing _else_ branch like this:
```
} else {
  return resolve();
}
```

File: gulpfile.js [hunk 8]
```diff
@@ -269,40 +272,46 @@ function resolveInstallerExePath(artifactType) {
...
+            // download only what can be included in offline installer
+            if (reqs[currentKey].bundle === 'yes') {
```
Comment (ppitonak) [inline gulpfile.js]: this branch is missing _else_ branch like this:
```
} else {
  return resolve();
}
```

**Comment by ppitonak** [comment]
@dgolovin apply changes from my comments and we are done :)

File: gulpfile.js (hunk 1 above)
Comment (dgolovin) [inline gulpfile.js]: I'll remove this, I was looking into prefetch task problems and wanted to be sure there is are no errors in Promises.

File: gulpfile.js (hunk 4 above)
Comment (dgolovin) [inline gulpfile.js]: It completes fast because all the problems @jrichter1 and @ppitonak pointed out above.

File: gulpfile.js (hunk 1 above)
Comment (dgolovin) [inline gulpfile.js]: This I really helpful in case of error inside promise. It at least print out error message, if you remove this callback there is no way to understand what is going on in case of error inside Promise instance.

**Comment by dgolovin** [comment]
I agree in general :) But the rule is not about lines it, it is more about cyclomatic complexity. You can literally remove line breaks from code to make it into 10 lines, but cyclomatic complexity will stay the same. And I can continue your list with "and not to have many return statements", because they increase complexity.

I'am going to fix all you suggestions and update PR.

File: gulpfile.js (hunk 1 above)
Comment (dgolovin) [inline gulpfile.js]: In our case to avoid problems like missing callbacks I would even re-throw the error, just to see the stacktrace.
```  JavaScript
process.on('uncaughtException', function(err) {
    if(err) {
      throw err;
    }
});
```
then I would see stacktrace for my missing callback 
```
C:\p\rh\dpi\gulpfile.js:50
      throw err;
      ^

TypeError: cb is not a function
    at createSHA256File (C:\p\rh\dpi\gulpfile.js:212:10)
    at WriteStream.<anonymous> (C:\p\rh\dpi\gulpfile.js:301:19)
    at emitNone (events.js:85:20)
    at WriteStream.emit (events.js:179:7)
    at finishMaybe (_stream_writable.js:478:14)
    at afterWrite (_stream_writable.js:356:3)
    at onwrite (_stream_writable.js:346:7)
    at WritableState.onwrite (_stream_writable.js:89:5)
    at fs.js:1951:5
    at FSReqWrap.wrapper [as oncomplete] (fs.js:667:5)
```

**Comment by ppitonak** [comment]
- [ ] `npm prune && rm -rf dist && rm -rf prefetch-dependencies && npm install`
- [ ] `npm start`
- [ ] `npm test`
- [ ] `npm run package-simple`
- [ ] `(( $(stat -c %s dist/win/*installer.exe) > 36000000 )) || echo 'file too small'`
- [ ] `(( $(stat -c %s dist/win/*installer.exe) < 40000000 )) || echo 'file too big'`
- [ ] start the application
- [ ] `npm run package-bundle`
- [ ] `(( $(stat -c %s dist/win/*installer.exe) > 1500000000 )) || echo 'file too small'`
- [ ] `(( $(stat -c %s dist/win/*installer.exe) < 1600000000 )) || echo 'file too big'`
- [ ] start the application
- [ ] `echo "aaa" > prefetch-dependencies/cdk.zip && npm run package-bundle`
- [ ] `echo "aaa" > prefetch-dependencies/cdk.zip && npm run dist`
- [ ] `npm run package-simple`
- [ ] `(( $(stat -c %s dist/win/*installer.exe) > 36000000 )) || echo 'file too small'`
- [ ] `(( $(stat -c %s dist/win/*installer.exe) < 40000000 )) || echo 'file too big'`

File: gulpfile.js [hunk 9]
```diff
@@ -173,67 +185,56 @@ gulp.task('create-sha256sum-of-exe', function(cb) {
...
+    }
+  }).on('end', function () {
+    hashstring = hash.digest('hex');
+    cb(hashstring);
+  });
 }
 
 // writes to {filename}.sha256, eg., 6441cde1821c93342e54474559dc6ff96d40baf39825a8cf57b9aad264093335 requirements.json
 function createSHA256File(filename, cb) {
+  !cb && cb();
```
Comment (ppitonak) [inline gulpfile.js]: What is this?

**Comment by ppitonak** [comment]
Merged to master. Thanks @nickboldt and @dgolovin

File: gulpfile.js (hunk 9 above)
Comment (dgolovin) [inline gulpfile.js]: I'll remove that, it just force failure before calculating sha if cb is not defined.
